webern/bottlerocket
# updog and migrator: signed migrations

**Issue number:**

TODO9XX
TODO91


**Description of changes:**

Implements all changes necessary for migrator to load a locally cached tuf repo and obtain migrations from there. See TODO9XX for a detailed explanation of the design.

One behavioral change about migrations. Previously we used a `sort` of the migration names to provide stability to their order of execution (when more than one migration exists for a single version pair). This is no longer necessary and the migrations will run in the order that they are listed in the manifest for a single version pair. The end-to-end migrator rest verifies this.

**Testing done:**

Added an end-to-end test of migrator.
Performed upgrade/downgrade testing as follows:
  * Create a build of these changes, give it version number 0.99.0
  * Create a build of these changes plus a couple of migrations, call it 0.99.1
  * Run 0.99.0, check that a pod runs, make a copy of the settings.
  * Upgrade to 0.99.1
  * Check that a pod runs, make a copy of the settings and verify that a migration created a "foo": "bar" setting.
  * Do a downgrade to 0.99.0
  * Check that a pod runs.
  * Make a copy of the settings and verify that "foo": "bar" was removed.

**TODO:**
  * Use pentacle instead of executing migrations from disk.
  * The end-to-end migrator test might be easier to understand if we generated the TUF repo on-the-fly instead of checking it into the git repository. It would also be much, easier to change. As soon as the `pubsys`/`tough` work progresses a bit, I will make the change.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.
